### PR TITLE
Added methods to convert Asn1TimeRef to a Unix time stamp.

### PR DESCRIPTION
**Pull request by illegalprime**
I have to verify a bunch of SSL certificates and I couldn't find another way to do it.

### CODE REVIEW

File: openssl/src/asn1.rs [hunk 1]
```diff
@@ -51,6 +51,25 @@ impl fmt::Display for Asn1TimeRef {
     }
 }
 
+impl Asn1TimeRef {
+    pub fn as_unix(&self) -> Result<i64, ErrorStack> {
```
Comment (sfackler) [inline openssl/src/asn1.rs]: I think it might make more sense to wrap these OpenSSL functions more directly.

File: openssl/src/asn1.rs [hunk 2]
```diff
@@ -51,6 +51,25 @@ impl fmt::Display for Asn1TimeRef {
...
+        let mut days = 0;
+        let mut seconds = 0;
+
+        unsafe {
+            let epoch = ffi::ASN1_TIME_set(ptr::null_mut(), 0);
```
Comment (sfackler) [inline openssl/src/asn1.rs]: This also leaks `epoch`.

Comment (illegalprime) [inline openssl/src/asn1.rs]: :+1:

File: openssl/src/asn1.rs (hunk 1 above)
Comment (illegalprime) [inline openssl/src/asn1.rs]: as in split it up into separate `set` and `diff` functions?

**Comment by illegalprime** [comment]
@sfackler some checks seem to be failing at the setup steps:

https://circleci.com/gh/sfackler/rust-openssl/674?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

**Comment by sfackler** [comment]
That failure is weird - I can repro it locally but that install was working fine as of 3 days ago... I'm poking around.

**Comment by sfackler** [comment]
Ah, you should be good to go if you rebase against current master.

**Comment by illegalprime** [comment]
yup that did it! Thanks!

File: openssl/src/asn1.rs [hunk 3]
```diff
@@ -51,6 +51,29 @@ impl fmt::Display for Asn1TimeRef {
...
 }
 
+impl Asn1TimeRef {
+    #[cfg(any(ossl102, ossl110))]
+    pub fn as_unix(&self) -> Result<time_t, ErrorStack> {
+        Asn1Time::from_unix(0)?.diff(Some(self))
+    }
+
+    #[cfg(any(ossl102, ossl110))]
+    pub fn diff(&self, compare: Option<&Self>) -> Result<time_t, ErrorStack> {
```
Comment (sfackler) [inline openssl/src/asn1.rs]: Can this return a `struct TimeDiff { days, secs }` instead? You can't generally assume days are the same length.

File: openssl/src/asn1.rs [hunk 4]
```diff
@@ -65,6 +88,16 @@ impl Asn1Time {
     pub fn days_from_now(days: u32) -> Result<Asn1Time, ErrorStack> {
         Asn1Time::from_period(days as c_long * 60 * 60 * 24)
     }
+
+    pub fn from_unix(time: time_t) -> Result<Self, ErrorStack> {
+        unsafe {
+            let asntime = ffi::ASN1_TIME_set(ptr::null_mut(), time);
```
Comment (sfackler) [inline openssl/src/asn1.rs]: This needs to call `ffi::init` first.

File: openssl/src/asn1.rs (hunk 3 above)
Comment (illegalprime) [inline openssl/src/asn1.rs]: Then `as_unix` would be difficult to implement, also a constant like this is used here: https://github.com/sfackler/rust-openssl/blob/master/openssl/src/asn1.rs#L66

File: openssl/src/asn1.rs (hunk 4 above)
Comment (illegalprime) [inline openssl/src/asn1.rs]: :+1:

File: openssl/src/asn1.rs (hunk 3 above)
Comment (sfackler) [inline openssl/src/asn1.rs]: Why would it be difficult to implement? This function would return exactly the same information as `ASN1_time_diff` does.

Comment (sfackler) [inline openssl/src/asn1.rs]: That constructor should probably not exist.

File: openssl/src/asn1.rs [hunk 5]
```diff
@@ -51,6 +57,32 @@ impl fmt::Display for Asn1TimeRef {
     }
 }
 
+impl Asn1TimeRef {
+    #[cfg(any(ossl102, ossl110))]
+    pub fn diff_from_epoch(&self) -> Result<TimeDiff, ErrorStack> {
+        Asn1Time::from_unix(0)?.diff(Some(self))
+    }
+
+    #[cfg(any(ossl102, ossl110))]
```
Comment (sfackler) [inline openssl/src/asn1.rs]: These should be flagged on features like here: https://github.com/sfackler/rust-openssl/blob/master/openssl/src/ssl/mod.rs#L1415

Comment (illegalprime) [reply openssl/src/asn1.rs]: I think the link broke, what do you mean?

Comment (sfackler) [reply openssl/src/asn1.rs]: https://github.com/sfackler/rust-openssl/blob/master/openssl/src/ssl/mod.rs#L1186-L1203

**Comment by clementsnahum** [comment]
Is there a reason this PR was closed? It looks like useful functionality.

**Comment by clementsnahum** [comment]
@sfackler if I were to resurrect this PR would you have any concerns about getting it merged? It looks as though you might be considering alternatives for comparing `Asn1Time` (https://github.com/sfackler/rust-openssl/issues/1157)

**Comment by sfackler** [comment]
Please do!

**Pull request closed without merging**